scripts/requirements-doc.txt

```diff
@@ -1,6 +1,6 @@
 # DOC: used to generate docs
 
-breathe>=4.23.0
+breathe>=4.23.0,<4.29.0
```

Comment:
Isn't there a !=4.29.0 we could use instead of '<', since the fix is posted and presumably will be in 4.30.0 or so?

Reply:
Assuming they fix the problem on 4.30.0 yes, we can. In general I prefer to do periodic manual upgrades of dependencies, I got hit too many times by these sort of problems... Zephyr should probably think about defining a policy on this aspect.

Reply:
>  In general I prefer to do periodic manual upgrades of dependencies, I got hit too many times by these sort of problems... Zephyr should probably think about defining a policy on this aspect.

I think that should be done with something like pipenv or a separate frozen requirements.txt; for the reasons we've discussed elsewhere, pinning versions in a requirements.txt that we ask people to install into their `--user` environments is IMO bad engineering.

Reply:
Lock files are probably a good solution to allow being permissive on requirements file while guaranteeing a stable CI. I used `pipenv` on the past, but there was a period when it was so unstable and not properly maintained I went back to plain `venv`. I haven't checked again if it's now stable enough for production use.

Being too permissive creates one problem: reproducible builds. If I go back to e.g. Zephyr 2.5.0 release in 2-3 years and I want to build the docs... err... good luck. The same will happen to all other Python based scripts on Zephyr, though.

An intermediate solution is to make use of https://www.python.org/dev/peps/pep-0440/#compatible-release version specifier, I use it on some projects with reasonable success.

Reply:
Sure, I was just throwing pipenv out there as an example; filtering `pip3 --freeze` output for a known working configuration and saving that in a separate requirements-ci.txt works too.

Compatible release specifiers would be a reasonable intermediate solution, I agree.